---
title: "02 - Prep normal reference signatures"
author: "Selin Jessa [[selin.jessa@mail.mcgill.ca](mailto:selin.jessa@mail.mcgill.ca)] and Bhavyaa Chandarana [[bhavyaa.chandarana@mail.mcgill.ca](mailto:bhavyaa.chandarana@mail.mcgill.ca)]"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: yes
    code_folding: show
    theme: flatly
    css: ../include/style.css
    toc: yes
    toc_depth: 4
    number_sections: true
    df_print: paged
    includes:
      before_body: ../include/header.html
      after_body:  ../include/footer.html
---

<!-- FRONT MATTER, insert configuration info -->

```{r header, echo = FALSE, results = 'asis', warning = FALSE}

# Index of the document
# ...determines name of the subfolder of `outputs` and `figures`
doc_id <- "02"

suppressMessages(library(here))

# Knit child document with header
res <- knitr::knit_child(here("include", "header.Rmd"), envir = environment(), quiet = TRUE)
cat(res, sep = '\n')

```

<!-- END OF FRONT MATTER -->


# Overview

Here we'll assemble a set of cell type signatures from the developing brain / adult brain,
building off the collection assembled for our
[Chen et al, Cell, 2020](https://www.cell.com/cell/fulltext/S0092-8674(20)31529-4)
study on G34R/V gliomas.

I will also add additional signatures for other brain datasets.

_NOTE_: This document re-uses a few functions for processing markers into lists
of gene signatures, `filter_markers()` and `signatures_to_list()` which are located
in `code/functions/scRNAseq.R`.


# Libraries

```{r libs, warning = FALSE, message = FALSE, echo = TRUE, cache = FALSE}

library(here)
library(magrittr)
library(tidyr)
library(dplyr)
library(readr)
library(readxl)
library(glue)
library(purrr)
library(ggplot2)
library(pheatmap)
library(Seurat)

source(here("include/style.R"))
source(here("code/functions/scRNAseq.R"))

ggplot2::theme_set(theme_min())

conflicted::conflicts_prefer(dplyr::filter)

```



# Gather signatures

Set function for conversion ensembl <-> symbol.

```{r}

signatures_to_list <- function(signatures, type = "ens") {
    
    if (type == "ens") {
        
        signatures %>% 
            select(Cluster, gene) %>%
            split(.$Cluster) %>%
            map(~ pull(.x, gene) %>% symbols2ensembl())
        
    } else if (type == "sym") {
        
        signatures %>% 
            select(Cluster, gene) %>%
            split(.$Cluster) %>%
            map(~ pull(.x, gene))
    }
}

symbols2ensembl <- function(genes) {
    
    load(here("data/singlecell/references_genome/biomaRt_hg_symbol_to_ens_lds.Rda"))
    genes_ens_lds %>% filter(HGNC.symbol %in% genes) %>% pull(Gene.stable.ID)
    
}

```

## Signatures compiled for Chen et al, 2020

In our analysis of H3G34R/V gliomas, we assembled signatures
from several studies on pre and postnatal forebrain, which we will re-use here ([GitHub link](https://fungenomics.github.io/G34-gliomas/bulk_transcriptome_epigenome/analysis/02-GSEA.html)). Copying over the comments from the Chen et al analysis:

> We obtained reference datasets for the forebrain across species and the lifespan:
Jessa et al, 2019; Nowakowski et al, 2017; Velmeshev et al, 2019
We also obtained two reference datasets capturing the adult SVZ:
Anderson et al, 2020; Mizrak et al, 2019


```{r load_signatures}

# laod gene signatures
signatures_ens <- readRDS(here("data/singlecell/references_normal/Chen_Cell_2020/signatures_ens.Rds"))
signatures_sym <- readRDS(here("data/singlecell/references_normal/Chen_Cell_2020//signatures_sym.Rds"))

# load signature annotation
cell_type_anno <- read_tsv(here("data/singlecell/references_normal/Chen_Cell_2020/signature_annotations.tsv"))

```

This set contains signatures from the our mouse forebrain atlas published in Jessa et al,
Nature Genetics, 2019. Since we have recently extended the timepoints covered by this atlas (Jessa et al, Nature Genetics, 2022), we will remove all the 2019 signatures from this collection. In the next section, we load in both 2019 & 2022 signatures.

```{r filter_sigs}

cell_type_anno_filt <- cell_type_anno %>% 
    filter(Dataset != "Jessa et al 2019")

signatures_ens <- signatures_ens[cell_type_anno_filt$Cluster]
signatures_sym <- signatures_sym[cell_type_anno_filt$Cluster]

```

## Jessa et al, Nature Genetics, 2022

Load the signatures for the extended atlas, combining timepoints published in 2019 and 2022:

```{r jessa_signatures}

# load signatures
jessa_signatures <- readRDS(here("data/singlecell/references_normal/Jessa_NatGenet_2022/joint_mouse_extended.signatures_ID_20210710.Rds"))

# subset to forebrain
jessa_signatures_ens <- jessa_signatures$hg_ens[grepl("^F-", names(jessa_signatures$hg_ens))]
jessa_signatures_sym <- jessa_signatures$hg_sym[grepl("^F-", names(jessa_signatures$hg_sym))]

# load annotation
jessa_cell_type_anno <- read_tsv(here("data/singlecell/references_normal/Jessa_NatGenet_2022/metadata_20210710_with_qc_v2.tsv")) %>%
    filter(Location == "Forebrain") %>% 
    filter(!grepl("EXCLUDE", Label)) %>% 
    filter(Label %in% names(jessa_signatures_ens)) %>% 
    select(Cluster = Label,
           Age = Timepoint,
           Cell_type = Level4_type) %>% 
    mutate(Dataset = "Jessa et al 2022",
           Sample = "Mouse forebrain")

# sanity check
nrow(jessa_cell_type_anno) == length(jessa_signatures_ens)
nrow(jessa_cell_type_anno) == length(jessa_signatures_sym)

```


## Aldinger et al, Nat Neurosci, 2021

```{r aldinger}

# load cell annotaiton
aldinger_anno <- read_xlsx(here("data/singlecell/references_normal/Aldinger_NatNeurosci_2021/data/Aldinger_SupplementaryTables.xlsx"), sheet = 8, skip = 2) %>% 
    dplyr::rename(Cluster = `Cluster-Cell Type Abbreviation`, Cell_type = `Cell Type`)

# load cluster markers
load(here("data/singlecell/references_normal/Aldinger_NatNeurosci_2021/aldinger_markers.Rda"))

# convert to signatures
aldinger_signatures <- aldinger_markers %>% 
    mutate(Cluster = paste0("Human fetal cerebellum ", cluster)) %>% 
    dplyr::rename(avg_logFC = avg_log2FC) %>% 
    filter_markers(n_top = 100)

sort(table(aldinger_signatures$Cluster))

aldinger_signatures_ens <- aldinger_signatures %>%
    signatures_to_list("ens")

aldinger_signatures_sym <- aldinger_signatures %>%
    signatures_to_list("sym")

aldinger_cell_type_anno  <- aldinger_anno %>%
    select(-3, -4, -5) %>% 
    mutate(Cluster = paste0("Human fetal cerebellum ", Cluster)) %>% 
    mutate(Dataset = "Aldinger et al 2021",
           Sample = "Human fetal cerebellum",
           Age = "9-21 PCW")

```


## Dong et al, Cancer Cell

This study profiled human fetal embryos and adrenal glands as a reference for 
studies of extra-cranial neuroblastoma.

```{r dong_et_al}

dong_embryo_markers <- read_xlsx(here("data/singlecell/references_normal/Dong_CancerCell_2020/Dong_TableS3.xlsx"), skip = 2) %>%
    dplyr::rename(Cluster = `Cell  type`, gene = Gene)

dong_sympatho_markers <- read_xlsx(here("data/singlecell/references_normal/Dong_CancerCell_2020/Dong_TableS4.xlsx"), skip = 2) %>%
    dplyr::rename(Cluster = Cell_type, gene = Gene)

dong_signatures <- bind_rows(dong_embryo_markers %>% mutate(Cluster = paste0("Human 4PCW ", Cluster)),
                             dong_sympatho_markers %>% mutate(Cluster = paste0("Human fetal adrenal ", Cluster))) %>%
    # take care of a common cluster name between datasets
    mutate(Cluster = ifelse(Cluster == "Human fetal adrenal SCPs", "Human fetal adrenal SCP", Cluster)) %>% 
    filter_markers(sp = "hg", n_top = 100)

table(dong_signatures$Cluster)

dong_signatures_ens <- dong_signatures %>% signatures_to_list("ens")
dong_signatures_sym <- dong_signatures %>% signatures_to_list("sym")

dong_cell_type_anno <- dong_signatures %>%
    distinct(Cluster) %>%
    mutate(Dataset = "Dong et al 2020",
           Age = "Human fetal",
           Cell_type = Cluster,
           Sample = case_when(
               grepl("4PCW", Cluster) ~ "Human 4PCW embryo neural lineages",
               grepl("adrenal", Cluster) ~ "Human 8-14PCW fetal adrenal glands"
           ))

```



## Jansky et al, Nature Genetics, 2021

This study also profiled fetal human adrenal glands and adrenal medulla for 
comparisons/study of neuroblastoma.

```{r jansky_et_al}

jansky_4 <- read_xlsx(here("data/singlecell/references_normal/Jansky_NatGenet_2021/Jansky_et_al_SupplementaryTables1-11.xlsx"), sheet = "Suppl. Table 4", skip = 1)
jansky_5 <- read_xlsx(here("data/singlecell/references_normal/Jansky_NatGenet_2021/Jansky_et_al_SupplementaryTables1-11.xlsx"), sheet = "Suppl. Table 5", skip = 1)

jansky_signatures <- bind_rows(jansky_4, jansky_5) %>% 
    dplyr::rename(Cluster = cluster,
                  avg_logFC = `average logFC`) %>% 
    filter_markers() %>%
    mutate(Cluster = paste0("Human fetal adrenal2 ", Cluster))

jansky_signatures_ens <- jansky_signatures %>% signatures_to_list("ens")
jansky_signatures_sym <- jansky_signatures %>% signatures_to_list("sym")

jansky_cell_type_anno <- jansky_signatures %>%
    distinct(Cluster) %>%
    mutate(Dataset = "Jansky et al 2021",
           Age = "Human fetal 7-17PCW",
           Cell_type = Cluster,
           Sample = "Human fetal adrenal gland")

```


## Kildisiute et al, Science Advances, 2021

This study also profiled human fetal adrenal glands as a reference for 
studies of extra-cranial neuroblastoma.

These markers were called with a custom algorithm by the authors, as described here
by the authors in their Methods:

> Using well-established marker genes of different cell types curated  from the literature (table S2), we assigned a cell type to each cluster.  Where two clusters were annotated as the same cell type, we merged  them together. As further confirmation of our annotation, we next  identified marker genes for each population algorithmically (table S6).  To do this, we used a method that uses the tf-idf metric to identify  genes specific to each population, as implemented in the “quickMarkers”  function in the SoupX R package (29). We further filtered genes to  include only those genes with a P value less than 0.01 after multiple  hypothesis correction (hypergeometric test).

Let's plot the markers produced in this study, comparing for each gene, the TF-IDF metric with the difference in detection rate of the gene in cells inside the cluster compared to all other cells. Each point is one gene.

```{r inspect_kildisiute, fig.width = 6, fig.height = 5}

kildisiute_markers <- read_xlsx(here("data/singlecell/references_normal/Kildisiute_SciAdv_2021/Kildisiute_Tables_S1_to_S12.xlsx"),
                                sheet = "TableS6.algorithmic_markers") %>%
    # filter out tumor datasets
    filter(dataset %in% c("adrenal_gland", "stringent_adrenal")) %>%
    mutate(pct_diff = geneFrequency - geneFrequencyOutsideCluster) %>%
    dplyr::rename(Cluster = cluster) %>%
    filter(Cluster != "Other") %>%
    mutate(Cluster = paste0("Human fetal adrenal ", Cluster))

# quite a large number of markers
table(kildisiute_markers$Cluster, kildisiute_markers$dataset)

# explore the dataset
kildisiute_markers %>%
    ggplot(aes(x = pct_diff, y = tfidf)) +
    geom_point(aes(colour = Cluster), alpha = 0.3) +
    geom_smooth(aes(colour = Cluster), method = "lm", se = FALSE) +
    xlab("Diff. in detection rate inside and outside cluster") +
    ylab("TF-IDF metric") +
    ggtitle("Before filtering")

```

This tells us that a high TF-IDF metric correlates with high/specific detection rate.
However, for some clusters, there's more variability along the X-axis, so I'll select markers based on the difference in detection rates (`pct_diff`), rather than the TF-IDF metric..

```{r kildisiute_sigs, fig.width = 6, fig.height = 5}

kildisiute_signatures <- kildisiute_markers %>%
    # filter out mitochondrial and ribosomal genes
    dplyr::filter(., !grepl("RPS|RPL|MRPS|MRPL|^MT-", gene)) %>%
    group_by(Cluster) %>%
    arrange(desc(pct_diff)) %>%
    # make sure they're unique
    distinct(gene, .keep_all = TRUE) %>%
    dplyr::slice(1:100) %>%
    ungroup()

# re-plot the same figure with th efiltered markers
kildisiute_signatures %>%
    ggplot(aes(x = pct_diff, y = tfidf)) +
    geom_point(aes(colour = Cluster), alpha = 0.3) +
    geom_smooth(aes(colour = Cluster), method = "lm", se = FALSE) +
    xlab("Diff. in detection rate inside and outside cluster") + 
    ylab("TF-IDF metric") +
    xlim(c(0, 1)) + ylim(c(0, 5)) +
    ggtitle("After filtering")

table(kildisiute_signatures$Cluster)

kildisiute_signatures_ens <- kildisiute_signatures %>% signatures_to_list("ens")
kildisiute_signatures_sym <- kildisiute_signatures %>% signatures_to_list("sym")

kildisiute_cell_type_anno <- kildisiute_signatures %>%
    distinct(Cluster) %>%
    mutate(Dataset = "Kildisiute et al 2021",
           Age = "Human fetal 8-21PCW",
           Cell_type = Cluster,
           Sample = "Human fetal adrenal glands")

```


## Shi et al, Science, 2021

Here we'll select markers from Shi et al, Science, 2021, which correspond to differentially expressed genes between different human fetal
ganglionic eminence progenitor clusters.

```{r load_shi}

shi_table_s6 <- read_xlsx(here("data/singlecell/references_normal/Shi_Science_2021/data/Shi2021_table_s6.xlsx"), skip = 1)

# how many genes per cluster?
table(shi_table_s6$cluster)
length(unique(shi_table_s6$gene))

shi_signatures_ens <- shi_table_s6 %>%
    select(cluster, gene) %>%
    mutate(cluster = paste0("Human fetal ", cluster)) %>%
    split(.$cluster) %>%
    map(~ pull(.x, gene) %>% symbols2ensembl())

shi_signatures_sym <- shi_table_s6 %>%
    select(cluster, gene) %>%
    mutate(cluster = paste0("Human fetal ", cluster)) %>%
    split(.$cluster) %>%
    map(~ pull(.x, gene))

signatures_ens2 <- c(signatures_ens, shi_signatures_ens)

shi_cell_type_anno  <- shi_table_s6 %>%
    distinct(cluster) %>%
    select(Cluster = cluster) %>%
    mutate(Cluster = paste0("Human fetal ", Cluster)) %>%
    mutate(Dataset = "Shi et al 2021", Sample = "Human fetal GE", Age = "Human fetal",
           Cell_type = case_when(
               grepl("C", Cluster) ~ "CGE progenitor",
               grepl("M", Cluster) ~ "MGE progenitor",
               grepl("L", Cluster) ~ "LGE progenitor"
           ))

```

## Van Bruggen et al, Dev Cell, 2022

This study focused on emergence of OPCs in the early human fetal brain.

```{r vanbruggen_markers}

vanbruggen_anno <- read_tsv(here("data/singlecell/references_normal/VanBruggen_DevCell_2022/cluster_annotation.tsv"))

```

We noted issues with gene naming in the supplementary tables provided with the manuscript,
so we re-called markers using `Seurat::FindAllMarkers()` and load them here:

```{r vanbruggen2}

load(here("data/singlecell/references_normal/VanBruggen_DevCell_2022/vanbruggen_markers2.Rda"))

scRNA_human_dev <- readRDS(here("data/singlecell/references_normal/VanBruggen_DevCell_2022/scRNA_human_dev.rds"))
vanbruggen_anno <- data.frame("Cluster" = names(sort(table(scRNA_human_dev$Clusters), decreasing = TRUE))) %>% 
    tibble::rowid_to_column(var = "Number")

vanbruggen_signatures <- vanbruggen_markers2 %>% 
    dplyr::rename(avg_logFC = avg_log2FC, Cluster = cluster) %>% 
    filter_markers(n_top = 100) %>% 
    left_join(vanbruggen_anno, by = "Cluster") %>% 
    mutate(Cluster = paste0("Human fetal ", Number, "-", Cluster))

sort(table(vanbruggen_signatures$Cluster))

vanbruggen_signatures_ens <- vanbruggen_signatures %>%
    signatures_to_list("ens")

vanbruggen_signatures_sym <- vanbruggen_signatures %>%
    signatures_to_list("sym")

vanbruggen_cell_type_anno  <- vanbruggen_signatures %>%
    distinct(Cluster) %>%
    mutate(Dataset = "VanBruggen et al 2021",
           Sample = "Human fetal forebrain",
           Age = "8-10GW",
           Cell_type = Cluster)

```


## Yu et al, Nature Neuroscience, 2021

This study also investigated human fetal ganglionic eminences and cortical interneuron development.

```{r yu_setup}

yu_cluster_markers <- read_xlsx(here("data/singlecell/references_normal/Yu_NatNeurosci_2021/Yu_NatNeuro_2021_Sup_tables.xlsx"),
                                sheet = 3, skip = 2) %>% 
    filter(!is.na(Cluster)) %>% 
    filter(!(Cluster %in% c("MG", "EC1", "EC2"))) %>% 
    mutate(Cluster = paste0("Fetal GE ", Cluster)) %>% 
    mutate(P_val = as.numeric(P_val), P_val_adj = as.numeric(P_val_adj))

yu_MGE_markers <- read_xlsx(here("data/singlecell/references_normal/Yu_NatNeurosci_2021/Yu_NatNeuro_2021_Sup_tables.xlsx"),
                            sheet = 11, skip = 2) %>% 
    mutate(Cluster = paste0("MGE-", Cluster))

yu_LGE_markers <- read_xlsx(here("data/singlecell/references_normal/Yu_NatNeurosci_2021/Yu_NatNeuro_2021_Sup_tables.xlsx"),
                            sheet = 12, skip = 2) %>% 
    mutate(Cluster = paste0("LGE-", Cluster)) %>% 
    mutate(P_val = as.numeric(P_val), P_val_adj = as.numeric(P_val_adj))

yu_CGE_markers <- read_xlsx(here("data/singlecell/references_normal/Yu_NatNeurosci_2021/Yu_NatNeuro_2021_Sup_tables.xlsx"),
                            sheet = 15, skip = 2) %>% 
    mutate(Cluster = paste0("CGE-", Cluster))

yu_markers <- bind_rows(yu_cluster_markers,
                        yu_MGE_markers,
                        yu_LGE_markers,
                        yu_CGE_markers) %>% 
    rename(gene = Gene, avg_logFC = Avg_logFC) %>% 
    mutate(Cluster = gsub("Fetal", "Human fetal", Cluster))

```

How many clusters, and how many markers per cluster?

```{r yu_n_clusterse}

unique(yu_markers$Cluster)
table(yu_markers$Cluster)

```

The imbalance in number of markers per cluster means that we should do some filtering to get them to roughly equal length:

```{r yu_signatures}

yu_signatures <- yu_markers %>% 
    filter_markers(n_top = 100)

sort(table(yu_signatures$Cluster))

yu_signatures_ens <- yu_signatures %>%
    signatures_to_list("ens")

yu_signatures_sym <- yu_signatures %>%
    signatures_to_list("sym")

yu_cell_type_anno  <- yu_signatures %>%
    distinct(Cluster) %>%
    mutate(Dataset = "Yu et al 2021",
           Sample = "Human fetal GE",
           Age = "Human fetal",
           Cell_type = Cluster)

```


## Compile all signatures

```{r compile_sigs, fig.height=4, fig.width=12, cache = FALSE}

cell_type_anno_all <- bind_rows(cell_type_anno_filt,
                                jessa_cell_type_anno,
                                shi_cell_type_anno,
                                yu_cell_type_anno,
                                vanbruggen_cell_type_anno,
                                dong_cell_type_anno,
                                kildisiute_cell_type_anno,
                                jansky_cell_type_anno,
                                aldinger_cell_type_anno) %>%
    mutate(Species = case_when(
        grepl("Velmeshev|Nowakowski|Yu|Shi|Dong|Kildisiute|Aldinger|VanBruggen|Jansky", Dataset) ~ "Human",
        TRUE ~ "Mouse"
    ))

cell_type_anno_all %>% distinct(Dataset, Species) %>% arrange(Species)

signatures_ens_all <- c(signatures_ens, jessa_signatures_ens, shi_signatures_ens, yu_signatures_ens,
                        dong_signatures_ens, kildisiute_signatures_ens,
                        vanbruggen_signatures_ens, jansky_signatures_ens,
                        aldinger_signatures_ens)
signatures_sym_all <- c(signatures_sym, jessa_signatures_sym, shi_signatures_sym, yu_signatures_sym,
                        dong_signatures_sym, kildisiute_signatures_sym,
                        vanbruggen_signatures_sym, jansky_signatures_sym,
                        aldinger_signatures_sym)

# QC: check length
sig_length <- map_dbl(signatures_ens_all, length) %>%
    tibble::enframe("Signature", "Length") %>%
    arrange(Length) %>%
    mutate(Signature = factor(Signature, levels = Signature))

sig_length %>% ggplot(aes(x = Signature, y = Length)) +
    geom_col() +
    rotate_x() +
    geom_hline(yintercept = 75, colour = "red") +
    theme(axis.text.x = element_text(size = rel(0.5)))

signatures_keep <- sig_length %>%
    filter(Length >= 75) %>%
    pull(Signature) %>%
    as.character()

# what is being dropped
sig_length %>%
    filter(Length < 75) %>%
    pull(Signature) %>% 
    as.character()

# total number of signatures
dim(cell_type_anno_all)

```

Since every study uses a slightly different cell type naming scheme, we will assign a harmonized class label for each cluster, for ease of interpretation:

```{r class, cache = FALSE}

cell_type_anno_all <- cell_type_anno_all %>% 
    mutate(Keep = ifelse(Cluster %in% signatures_keep, TRUE, FALSE)) %>% 
    mutate(Class = case_when(
        grepl("-P|prolif|TOP2A|Cycling|div", Cluster) & !grepl("Pericytes", Cluster) ~ "Proliferating progenitors",
        grepl("RG|[Rr]adial", Cluster) &
            !grepl("NRGN", Cluster) &
            !grepl("[Oo]lig", Cluster) ~ "Radial glia",
        grepl("NSC", Cluster) ~ "Neural stem cell",
        grepl("OPC|Oligodendrocyte progenitor cell|Oligodendrocyte precurso|COP", Cluster) ~ "Oligodendrocyte precursors",
        grepl("NFOL|MOL|[Oo]ligo|OL", Cluster) ~ "Oligodendrocytes",
        grepl("ASEP|EPEN|[Ee]pendymal|[Cc]horoid|CPLX", Cluster) ~ "Ependymal",
        grepl("AST|Ast|Astr|BG", Cluster) ~ "Astrocytes",
        grepl("MNG|[Mm]eninges|VLM", Cluster) ~ "Meninges",
        grepl("Mesenchym", Cluster) ~ "Mesenchyme",
        grepl("PERI|[Pp]ericyte|[Mm]ural|Vascular|ENDO|[Ee]ndo|[Ff]ibroblast|VSMC", Cluster) ~ "Mural",
        grepl("Muscle|Myo", Cluster) ~ "Muscle",
        grepl("Glia|GLIP|[Gg]lioblast", Cluster) ~ "Glial progenitors",
        grepl("ERY|Immune|MGL|MAC|Microgl|Macro|Leuko|Erythro| T ", Cluster) ~ "Immune",
        grepl("SCP|SCHW|[Cc]hromaffin|Bridge|[Ss]ympatho|Adrenal|NCC|Melano", Cluster) ~ "Neural crest lineages",
        grepl("EN|EMX|EXN|ExN|[Ee]xcit|nEN|Pyram", Cluster) & !grepl("ENC", Cluster) ~ "Excitatory neurons",
        grepl("Human fetal GE [pP]", Cluster) ~ "GE neural precursors",
        grepl("GE|GABA|nIN|MGIN|MGE|CGE|LGE|SST|PV|[Ii]nhib|CIN|PV|SST|VIP|SV2C|Somato|Parv|pC|pM|pL|Inter", Cluster) &
            !grepl("[Mm]idbrain|[Hh]indbrain", Cluster) ~ "GE/Inhibitory neurons",
        grepl("EXIP|NEURP|IP|[Ii]ntermediate|prog|Neurogenic", Cluster) &
            !grepl("VIP", Cluster) & !grepl("Human fetal progenitor:", Cluster) ~ "Neuronal progenitors",
        grepl("SPN|Str|STR|SMSN|Striatum", Cluster) ~ "Striatal neurons",
        grepl("GN|iCN|INs|MLI|Moto|MN|NRGN|[Nn]eu|MFN|CJRZN|UBC|GABAN|NEUR|THLN|SN|INH|SRN|ACHN", Cluster) & !grepl("SMSN", Cluster) ~ "Other neurons",
        grepl("Human fetal cortex:|Human fetal progenitor:|Human fetal CS12-13:", Cluster) ~ "Unlabelled",
        TRUE ~ "Other"))

save(cell_type_anno_all, signatures_keep, signatures_sym_all, signatures_ens_all, signatures_keep,
     file = glue("{out}/signatures.Rda"))

```

# Prepare supplementary table

Here, we'll prepare a supplementary table containing all the reference gene signatures
and their associated cell types/datasets from which they originate.


```{r table_ref_sigs, cache = FALSE}

# collapse signatures into strings
sigs_sym_df <- imap_dfr(signatures_sym_all,
                        ~ data.frame(Cluster = .y, Signature_sym = glue_collapse(.x, sep = ",")))

sigs_ens_df <- imap_dfr(signatures_ens_all,
                        ~ data.frame(Cluster = .y, Signature_ENS = glue_collapse(.x, sep = ",")))

# join with cell type annotation
TABLE_ref_sigs <- cell_type_anno_all %>% 
    left_join(sigs_sym_df) %>% 
    left_join(sigs_ens_df) %>%
    # filter to kept signatures
    filter(Keep) %>% 
    select(-Keep) %>%
    mutate(Cell_type = ifelse(is.na(Cell_type), Cluster, Cell_type))

rr_write_tsv(TABLE_ref_sigs,
             glue("{out}/TABLE_reference_signatures.tsv"),
             "Cell type specific reference signatures and associated metadata and dataset info")

```


# Compare signatures

We have assembled signatures from two species, from multiple marker-identification approaches, and from multiple studies. As a sanity check, we compute pairwise Jaccard Index for overlap in genes between the signatures, in order to confirm that similar cell types do have similar gene signatures.

```{r signature_jaccard, cache = FALSE}

# calculate Jaccard index as follows:
# |X and Y| / |X or Y|
jaccard <- function(x, y) length(base::intersect(x, y)) / length(base::union(x, y))

signatures_overlap <- sapply(signatures_sym_all[signatures_keep],
                             function(x) sapply(signatures_sym_all[signatures_keep],
                                                function(y) jaccard(x, y)))

anno <- cell_type_anno_all %>% select(Cluster, Dataset, Species, Class) %>% tibble::column_to_rownames(var = "Cluster")

palette_dataset <- set_qual_pal(length(unique(anno$Dataset)))
names(palette_dataset) <- unique(anno$Dataset)

hm_jaccard <- pheatmap(signatures_overlap,
                       color             = custom_magma,
                       border_color      = NA,
                       treeheight_row    = 50,
                       treeheight_col    = 0,
                       cutree_rows       = 35,
                       cutree_cols       = 35,
                       annotation_row    = anno,
                       annotation_col    = anno,
                       annotation_colors = list("Dataset" = palette_dataset,
                                                "Species" = c("Mouse" = "gray90", "Human" = "black"),
                                                "Class"   = palette_class),
                       cellwidth         = 3,
                       cellheight        = 3,
                       fontsize_row      = 4,
                       fontsize_col      = 4,
                       filename          = glue("{figout}/signature_jaccard.png"))

# save heatmap to extract clustering order
save(hm_jaccard, file = glue("{out}/hm_jaccard.Rda"))

knitr::include_graphics(glue("{figout}/signature_jaccard.png"))

```

- We confirm that there is coherence across datasets, with major clusters for microglia, endothelial other vascular/non-neuroetodermal cells, OPCs, astro-related cells, cycling cells, different types of neurons
- This suggests that the signatures are informative and capturing cell type-specific rather that simply dataset-specific information


# Stats

```{r signature_stats, cache = FALSE}

cell_type_stats <- cell_type_anno_all %>%
    filter(Cluster %in% signatures_keep) %>% 
    mutate(Sample = case_when(
        Dataset == "Jessa et al 2019" & grepl("Forebrain", Sample) ~ "Mouse forebrain",
        Dataset == "Jessa et al 2019" & grepl("Pons", Sample) ~ "Mouse pons",
        TRUE ~ Sample
    )) %>% 
    mutate(Age = case_when(
        Dataset == "Jessa et al 2019" & grepl("forebrain", Sample) ~ "Mouse E12-P6",
        Dataset == "Jessa et al 2019" & grepl("pons", Sample) ~ "Mouse E12-P6",
        TRUE ~ Age
    )) %>% 
    group_by(Sample, Age, Dataset) %>% count() %>% arrange(Dataset)

sum(cell_type_stats$n)

kableExtra::kable(cell_type_stats)

```

```{r signature_stats_barplot, fig.width = 5, fig.height = 6, cache = FALSE}

cell_type_stats %>%
    tibble::rowid_to_column(var = "Order") %>% 
    ggplot(aes(x = Order, y = n)) +
    geom_col(aes(fill = Dataset), width = 0.5) +
    scale_fill_manual(values = palette_dataset) +
    coord_flip()

```



<!-- END MATTER, insert reproducibility info -->

```{r footer, echo = FALSE, results = 'asis', warning = FALSE, cache = FALSE}

# Knit child document with header
res <- knitr::knit_child(here("include", "footer.Rmd"), envir = environment(), quiet = TRUE)
cat(res, sep = '\n')

```

<!-- END OF END MATTER -->
